thread_primitives.h

```diff
@@ -28,65 +28,119 @@
 #ifndef THREAD_PRIMITIVES_H
 #define THREAD_PRIMITIVES_H
 
+#include <condition_variable>
+#include <functional>
+#include <future>
+#include <memory>
+#include <mutex>
+#include <thread>
+
 class ConditionVariablePrivate;
 class MutexPrivate;
 
 class Mutex {
-  public:
-    Mutex();
-	~Mutex();
-    void lock();
-    bool tryLock();
-    void unlock();
-
-    MutexPrivate* impl_;
+public:
+  Mutex() = default;
+  void lock() { mutex_.lock(); }
+  bool tryLock() { return mutex_.try_lock(); }
+  void unlock() { mutex_.unlock(); }
+
+  std::mutex mutex_;
 };
 
 //! Aquire a lock on the mutex passed to the constructor.
 //! The lock is maintained while the ScopedLock object is alive.
 //! The lock is released uppon destruction.
 class ScopedLock {
-  public:
-    ScopedLock(Mutex* mutex) : mutex_(mutex) { mutex->lock(); }
-    ~ScopedLock() { mutex_->unlock(); }
-
-  private:
-    ScopedLock(const ScopedLock&) { }
-    ScopedLock& operator = (const ScopedLock&) { return *this; }
-    Mutex* mutex_;
+public:
+  ScopedLock(Mutex *mutex) : mutex_(mutex) { mutex->lock(); }
```

Comment:
I would prefer to have style changes in a separate PR. No need to revert those.

Reply:
yes that was my automatic clang-format :/

Reply:
- [ ] make sure clang-format is not doing magic tricks :tophat: